pages/en/lb4/Testing-Your-Extensions.md

```diff
@@ -0,0 +1,303 @@
+
+### Manual Setup - Using Mocha
+- Install `mocha` by running `npm i --save-dev mocha`. This will save the package in
+`package.json` as well.
+- In `package.json` add under `scripts` the following:
+`test: npm run build && mocha --recursive ./dist/test`
+
+## Types of tests
+
+A comprehensive test suite tests many aspects of your code. It is encouraged to
+test as many aspects of your application from a variety of perspectives to
+ensure correctness, integration and ensuring future compatibility. To accomplish
+this you should write unit, integration and acceptance tests.
+
+You may use any development methodology you want to write your extension, the
+important thing is to have an automated test suite.
+- **Traditional development:** write the code first and then write the tests
+- **Test-driven development:** write tests first, see them fail, then write the
+code to pass the tests
+
+### Unit Tests
+
+A unit test checks the smallest unit of code possible, in this case a function.
+This ensures variable and state changes by outside actors don't affect the
+results. [Test doubles](https://en.wikipedia.org/wiki/Test_double) should be
+used to substitue function dependencies. You can learn more about test doubles
+and Unit testing [here: Testing your Application: Unit testing](http://loopback.io/doc/en/lb4/Testing-your-application.html#unit-testing).
+
+#### Controllers
+
+At its core, a controller is a simple class that is responsible for related
+actions on an object. Unit testing a controller in an extension is the same as
+unit testing a controller for an application.
+
+To test a controller you want to instantiate a new instance of your controller
+class and test a function, providing a test double for constructor arguments as
+needed.
+
+**`src/controllers/ping.controller.ts`**
+```ts
+export class PingController {
+  @get('/ping')
+  ping(msg?: string) {
+    return `You pinged with ${msg}`;
+  }
+}
+```
+
+**`test/unit/controllers/ping.controller.ts`**
+```ts
+import {PingController} from '../../..';
+import {expect} from '@loopback/testlab';
+
+describe('PingController() unit', () => {
+  it('pings with no input', () => {
+    const controller = new PingController();
+    const result = controller.ping();
+    expect(result).to.equal('You pinged with undefined');
+  });
+
+  it('pings with msg \'hello\'', () => {
+    const controller = new PingController();
+    const result = controller.ping('hello');
+    expect(result).to.equal('You pinged with hello');
+  });
+});
+```
+
+You can find a more advanced example on testing controllers in [Unit test your Controllers](http://loopback.io/doc/en/lb4/Testing-your-application.html#unit-test-your-controllers).
+
+#### Decorators
+
+The recommended usage of a decorator is to store metadata about a class or a
+class method. The decorator implementation will usually also provide a function
+to retrieve the metadata stored by it based on the className and method name.
```

Comment:
Instead of "based on the className and method name", how about "based on the class name and method name"?